core/src/main/java/com/datastrato/gravitino/storage/kv/KvEntityStore.java

```diff
@@ -355,6 +371,7 @@ public boolean delete(NameIdentifier ident, EntityType entityType, boolean casca
           List<byte[]> subEntityPrefix = getSubEntitiesPrefix(ident, entityType);
           if (subEntityPrefix.isEmpty()) {
```

Comment:
I don't think we need to add `deleteUserEntitiesIfNecessary` here, subEntityPrefix is empty when the entity to delete is a table, a fileset or entities in the leaf nodes. 

Instead, you may need to add some logic here:
![image](https://github.com/datastrato/gravitino/assets/15794564/60c7c72e-a753-4781-b319-c05ea3b671b3)

Reply:
> I don't think we need to add `deleteUserEntitiesIfNecessary` here, subEntityPrefix is empty when the entity to delete is a table, a fileset or entities in the leaf nodes.
> 
> Instead, you may need to add some logic here: ![image](https://private-user-images.githubusercontent.com/15794564/313574245-60c7c72e-a753-4781-b319-c05ea3b671b3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA3NDUyNzQsIm5iZiI6MTcxMDc0NDk3NCwicGF0aCI6Ii8xNTc5NDU2NC8zMTM1NzQyNDUtNjBjN2M3MmUtYTc1My00NzgxLWIzMTktYzA1ZWEzYjY3MWIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE4VDA2NTYxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5YzA2YzU5MzI3NDY0YmQ1YjI2ZjE5ZGY1N2JiOTNlMThjNTc0ODE0OGNjYjk2NmIxZDEzYjc5MGM0YzViMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bhjfqGQBuiq2ZO3OdTM_vlbVkw04QobjxSyLRSjPbfo)

I don't think so. The deletion of users don't follow the cascade rules. So I don't add it in this place.

Reply:
It's not so proper to add the logic here 
![image](https://github.com/datastrato/gravitino/assets/15794564/4334bf5f-9cfc-4684-bce8-73bc344ba948)

If the Entity type is Metalake, the value of subEntityPrefix will not be empty, so method `            deleteUserEntitiesIfNecessary(ident, entityType);` will never take effect actually as you add the check

```java
  void deleteUserEntitiesIfNecessary(NameIdentifier ident, EntityType type) throws IOException {
    if (type != METALAKE) {
      return;
    }
}
```

Reply:
Why won't this subEntityPrefix be empty? I don't include the users in this subEntityPrefix. It may be empty.

Reply:
`deleteUserEntitiesIfNecessary` onlys works when the entity type is `Metalake` and when the entity type is `metalake` , `subEntityPrefix` is NOT empty.

Reply:
> `deleteUserEntitiesIfNecessary` onlys works when the entity type is `Metalake` and when the entity type is `metalake` , `subEntityPrefix` is NOT empty.

Why when the entity type is metalake , subEntityPrefix is NOT empty? I don't second it.

Reply:
Please see the picture here:
https://github.com/datastrato/gravitino/pull/2481#discussion_r1527929699

Reply:
> Please see the picture here: [#2481 (comment)](https://github.com/datastrato/gravitino/pull/2481#discussion_r1527929699)

I don't add USER entity there. User entity won't block the deletion.